simpleviewinc/keg-cli
# Update expo and node

**Ticket**: [ZEN-526](https://jira.simpleviewtools.com/browse/ZEN-526)

## Context
* Expo has been breaking lately in the keg, preventing our apps from compiling
  * This is due to some issue with Expo that requires either deleting `.expo` or upgrading to the latest
  * see: https://github.com/expo/expo-cli/issues/2951
* We can fix this manually each time it arises
  * We should update the keg's version of expo since it is now behind and causing issues
* Keg's expo-cli
  * Current version: 3.28.5
  * Latest version: 4.0.17
* Expo sdk
  * Current version: 39
  * Latest version: 40

## Goal

* Update expo-cli and sdk versions for keg repos and taps
* Update docker images to include expo updates
  * Fix sharp-cli issue

## Updates
* Updated workflows to use node 14
* Update the sharp-cli version to be it's own ENV named `SHARP_CLI_VERSION`
* Updated keg-base docker image to install python, to allow installing the sharp-cli
* Updated `package.json` dependencies 
* `build` && `push` tasks
  * Fixed issue with from option not working properly 
* `src/tasks/hub/dependencies.js`
  * Updated task to allow updating the dependencies for all keg-hub repos
  * This will update the package.json file of all repos with a matching dependency
    * If uses the latest version found amoung all repos
  * It does **NOT** run any git commands
    * There may be cases where we don't want to update a dependency version for a repo
    * The changes are expected to be reviewed and tested before being committed to the repo

## Testing

**Setup**
* Pull this branch => `keg cli && keg pr 44`

**Test 1**
* Switch to the `develop` branch of the `keg-hub` repo
  * Run the command `keg hub deps`
    * It should run and display a list of mis-matched dependencies
    * It should **NOT** do any updates
  * Next run the command `keg hub deps --update`
    * It should run, and update the package.json of all mismatched dependencies
    * Open keg-hub in your editor and validate that each repos `package.json` was updated correctly
    * Reset all changes => `keg hub && keg reset -f`

**Test 2**
* Switch to the `ZEN-526-upgrade-expo` branch for the `keg-hub` repo
  * `keg hub && keg pr 144`
* Run the command `keg hub deps` again
  * It should print the text `No mismatched dependency versions found!`
  * This branch already has all the dependencies updated
  * Ensure no-mismatched dependencies are found
* Next run the command `keg hub deps --update`
  * Ensure it should prints out the same message as above =>  `No mismatched dependency versions found!`
  * Ensure no-mismatched dependencies are updated in the package.json files

**Test 3**
* Build the base image **with a tag** 
  * Run the command `keg base build --tag <replace-with-custom-tag>`
    * **IMPORTANT** - Be sure to use your own custom test tag
    * If you don't it will use the `master` tag which will overwrite you local master version
  * Ensure the base image build successfully

**Test 4**
  * Build keg-core passing in the `tag` and `from` options as follows
    * `tag` - Can be anything you want it to be i.e. `--tag <replace-with-custom-tag>`
      * **IMPORTANT** - Be sure to use your own custom test tag
    * `from` - Should be then name of the custom tag used when building `keg-base` in **Test 3**
  * The command should look something like this
    * `keg core build --tag core-custom-tag --from keg-base:base-custom-tag`
  * Ensure the keg-core image build successfully
  * Ensure it builds from the correct base image and tag
    * [Looks like similar to this](http://snpy.in/1cOrGh)
      * See the 6th blue line has the text `keg-base:upgrade-expo-sdk`
        * `upgrade-expo-sdk` is the custom tag I used when building my base image
      * This means it's using the `keg-base:upgrade-expo-sdk` image and not `keg-base:master`
  * Run the build command again, this time passing the `push` option
    * The build should be very quick, because it's cached
    * It should then push your image to the provider
      * **IMPORTANT** - If you didn't use a custom tag here, it will overwrite the base:master image and _I will be sad :(_
    * The command should look something like this
      * `keg core build --tag core-custom-tag --from keg-base:base-custom-tag --push`
    * Ensure it builds and pushes the **correct** `keg-core` image and tag to the provider

**Test 5**
  * Push keg-core passing in the `tag`, `from` and `build` options
  * It should basically do the exact same thing as in **Test 4**, but this time from the `push` task
  * The command should look something like this
    * `keg core push --tag upgrade-expo-sdk --from keg-base:upgrade-expo-sdk --build`
    * **IMPORTANT** - Be sure to use your own custom test tag
    * The build should be very quick, because it's cached
    * The push should be very quick because it's already been pushed to the provider
  * Ensure it builds and pushes the **correct** `keg-core` image and tag to the provider

**Test 6**
  * Build the tap image passing in the `tag`, `from` and `push` options
  * It should basically run the exact same command as in **Test 4**
  * The command should look something like this
    * `keg tap build --tag tap-custom-tag --from keg-base:base-custom-tag --push`
    * **IMPORTANT** - Be sure to use your own custom test tag
    * The build should be very quick, because it's cached
    * The push should be very quick because it's already been pushed to the provider

**Test 7**
  * Build a linked tap passing in the `tag`, `from` and `push` options
  * The command should look something like this
    * `keg evf build --tag evf-custom-tag --from tap:tap-custom-tag --push`
  * **IMPORTANT** - Linked taps should build from the `tap` image **NOT** the `base` image
    * So ensure your `--from` points to the `tap:<your-custom-tag>` image built in **Test 6**
  * Start EVF tap => `keg evf start --from evf:evf-custom-tag`
    * Ensure that tap starts without issue
      * You should not see ant sharp-cli warnings
      * You should be able to access the tap in the browser
    * In another tab attach to the container => `keg evf att`
    * Once attached,
      * Validate the expo version => `expo -V`
        * Should print `4.3.0`
      * Validate the node version => `node -v`
        * Should print `v14.16.0`
    * [Looks like this](http://snpy.in/OM3v2M)

**Test 8**
  * Run all the same commands in **Test 7** but switch out the linked tap for another one
    * If you used evf, you could try `keg-herkin` or `re-theme`
    * They should all work the same

> Notes
> Just figured I would bring up a few things
> The test steps above illustrate the benefits of building applications with the **Keg**
> As seen above, **NO Changes** are required by to any of the taps
> Yet they all receive the benefits of the updates to the `keg-cli` and `keg-hub repos`
> At first it may be hard to understand the design decisions of the **Keg**
> But as more taps are built and updates or changes are required;
> it becomes more clear how the benefits of the **Keg** can be realized
>
> Currently it's mostly just the Eventsforce Tap,
> but as we add more taps the benefits will be compounded.
> Hours of developer time will be saved, and support work will be easier
> We'll remove the need for separate build machines, one-off deploy scripts and more
> There's a bright future ahead :)
